python/lsst/skymap/tractBuilder.py

```diff
@@ -0,0 +1,378 @@
+        the given patch within this tract.
+
+        Parameters
+        ----------
+        patchInfo : `lsst.skymap.PatchInfo`
+
+        Returns
+        -------
+        sequentialIndex : `int`
+        """
+        index = patchInfo.getIndex()
+        return self.getSequentialPatchIndexFromPair(index)
+
+    def getSequentialPatchIndexFromPair(self, index):
+        """Return a single integer that uniquely identifies
+        the patch index within the tract.
+
+        Parameters
+        ----------
+        index : `lsst.skymap.Index2D` or `Iterable` [`int`, `int`]
+
+        Returns
+        -------
+        sequentialIndex : `int`
+        """
+        if isinstance(index, Index2D):
+            _index = index
+        else:
+            if not isinstance(index, Iterable):
+                raise ValueError("Input index is not an iterable.")
+            if len(index) != 2:
+                raise ValueError("Input index does not have two values.")
+            _index = Index2D(*index)
+        nx, ny = self._numPatches
+        return nx*_index.y + _index.x
+
+    def getPatchIndexPair(self, sequentialIndex):
+        """Convert sequential index into patch index (x,y) pair.
+
+        Parameters
+        ----------
+        sequentialIndex : `int`
+
+        Returns
+        -------
+        x, y : `lsst.skymap.Index2D`
+        """
+        nx, ny = self._numPatches
+        x = sequentialIndex % nx
+        y = sequentialIndex // nx
+        return Index2D(x=x, y=y)
+
+    @abc.abstractmethod
+    def getPackedConfig(self, config):
+        """Get a packed config suitable for using in a sha1.
+
+        Parameters
+        ----------
+        config : `lsst.skymap.BaseTractBuilderConfig`
+
+        Returns
+        -------
+        configPacked : `bytes`
+        """
+        raise NotImplementedError("Must be implemented by a subclass")
+
+
+class LegacyTractBuilderConfig(BaseTractBuilderConfig):
+    patchInnerDimensions = pexConfig.ListField(
+        doc="dimensions of inner region of patches (x,y pixels)",
+        dtype=int,
+        length=2,
+        default=(4000, 4000),
+    )
+    patchBorder = pexConfig.Field(
+        doc="border between patch inner and outer bbox (pixels)",
+        dtype=int,
+        default=100,
+    )
+
+    def validate(self):
+        if len(self.patchInnerDimensions) != 2:
+            raise ValueError("patchInnerDimensions must be 2 ints.")
+
+
+class LegacyTractBuilder(BaseTractBuilder):
+    ConfigClass = LegacyTractBuilderConfig
+
+    def __init__(self, config):
+        super().__init__(config)
+
+        self._patchInnerDimensions = geom.Extent2I(*(val
+                                                     for val in config.patchInnerDimensions))
+        self._patchBorder = config.patchBorder
+        self._initialized = False
+
+    def getPatchInfo(self, index, tractWcs):
+        # This should always be initialized
+        assert self._initialized
```

Comment:
Should this raise an AssertionError wihout any helpful message, or should we explicitly raise an exception? I understand `assert` statement are removed from the production builds, but during the development stage I don't think this is going to help and shouldn't be a significant overhead in production either.